File: crates/wormhole-cli/src/commands/submit.rs
use clap::Args;
use wormhole_sdk::chains::evm::{self, SubmitParams, DEVNET_FROM, DEVNET_RPC};

use crate::output;

/// Submit a VAA to an EVM chain.
#[derive(Debug, Args)]
pub struct SubmitArgs {
    /// Hex or base64-encoded VAA to submit.
    pub vaa: String,
    /// Target EVM chain name (e.g. ethereum).
    #[arg(short = 'c', long)]
    pub chain: Option<String>,
    /// Target contract address (overrides default).
    #[arg(short = 'a', long)]
    pub contract_address: Option<String>,
    /// RPC endpoint URL (overrides default).
    #[arg(long)]
    pub rpc: Option<String>,
    /// `from` address for devnet (unlocked account).  Defaults to Ganache account 0.
    #[arg(long, env = "WORMHOLE_EVM_FROM")]
    pub from: Option<String>,
    /// Secp256k1 private key hex for signed submission (testnet/mainnet).
    #[arg(long, env = "WORMHOLE_EVM_KEY")]
    pub evm_key: Option<String>,
}

/// Run the `worm submit` command.
///
/// Parses the VAA, detects its module, and submits it to the appropriate
/// Wormhole contract on the target chain.  Prints the transaction hash as JSON.
///
/// # Errors
///
/// Returns an error if the VAA cannot be parsed, the contract address cannot be
/// determined, or the RPC call fails.
pub async fn run(args: &SubmitArgs) -> anyhow::Result<()> {
    let rpc_url = args.rpc.as_deref().unwrap_or(DEVNET_RPC);
    let from = args.from.as_deref().or(Some(DEVNET_FROM));

    let contract_address = resolve_contract_address(args)?;

    let tx_hash = evm::submit_vaa(SubmitParams {
        vaa_hex: &args.vaa,
        rpc_url,
        contract_address: &contract_address,
        from_address: from,
        evm_key: args.evm_key.as_deref(),
    })
    .await
    .map_err(|e| anyhow::anyhow!("submit failed: {e}"))?;

    output::print_json(&tx_hash)
}

/// Resolve the target contract address from CLI args or env vars.
fn resolve_contract_address(args: &SubmitArgs) -> anyhow::Result<String> {
    if let Some(addr) = &args.contract_address {
        return Ok(addr.clone());
    }

    // Try env vars for known modules
    if let Ok(addr) = std::env::var("WORMHOLE_CORE_ADDRESS") {
        return Ok(addr);
    }

    // Default: devnet core bridge
    Ok(evm::DEVNET_CORE.to_string())
}

File: packages/cli/src/commands/submit.ts
import type { Command } from 'commander';
import { parseVaa, getChainByName } from '@worm-tool/sdk';
import { loadConfig } from '../config.js';
import { createEvmChain } from '../providers/evm.js';
import { printJson, printError } from '../output.js';
import { abiEncodeBytes } from './_vaa-utils.js';

export function registerSubmitCommand(program: Command): void {
  program
    .command('submit <vaa>')
    .description('Submit a signed VAA to a Wormhole contract on an EVM chain')
    .requiredOption('--chain <name>', 'Target EVM chain name (e.g. ethereum)')
    .requiredOption('--selector <hex>', '4-byte function selector, e.g. 0x5cb8cae2 (submitContractUpgrade), 0xc6878519 (completeTransfer)')
    .option('--contract <address>', 'Target contract address (overrides default core bridge)')
    .action(async (vaa: string, opts: { chain: string; selector: string; contract?: string }) => {
      try {
        const config = loadConfig();
        const parsed = parseVaa(vaa);

        const chain = createEvmChain(opts.chain, config);

        const contract = opts.contract ?? (() => {
          const entry = getChainByName(opts.chain);
          if (!entry?.wormholeCore) throw new Error(`No known core bridge for ${opts.chain} — provide --contract`);
          return entry.wormholeCore;
        })();

        const selector = opts.selector.startsWith('0x') ? opts.selector : '0x' + opts.selector;
        const rawHex = vaa.startsWith('0x') ? vaa : '0x' + vaa;
        const data = (selector + abiEncodeBytes(rawHex).slice(2)) as `0x${string}`;

        const receipt = await chain.sendTransaction(contract, data);
        printJson({ txHash: receipt.txHash, success: receipt.success, chain: opts.chain, sequence: parsed.sequence.toString() });
      } catch (err) {
        printError('submit failed', err);
        process.exit(1);
      }
    });
}
